javalib/src/main/scala/java/lang/StrictMath.scala

```diff
@@ -0,0 +1,187 @@
+ * software is freely granted, provided that this notice
+ * is preserved.
+ * ====================================================
+ */
+
+package java.lang
+
+object StrictMath {
+
+  // Ported from https://www.netlib.org/fdlibm/s_floor.c
+  def floor(x: scala.Double): scala.Double = {
+    val doubleBits = Double.doubleToLongBits(x)
+
+    var highWord = (doubleBits >>> 32).toInt  // sign(1) + exponent(11) + mantissa_high(20)
+    var lowWord = doubleBits.toInt            // mantissa_low(32)
+    val exponent = ((highWord >> 20) & 0x7ff) - 1023
+
+    @inline def reconstructDouble(highWord: Int, lowWord: Int): scala.Double = {
+      val bits = (highWord.toLong << 32) | (lowWord.toLong & 0xffffffffL)
+      Double.longBitsToDouble(bits)
+    }
+
+    if (exponent < 20) { // Case 1: decimal point is withing the high word.
+      if (exponent < 0) { // Case 1a: |x| < 1
+        /* Note: In the original C implementation, code block below was surrounded by
+         * "if (huge + x > 0.0)" (huge = 1.0e300),
+         * which was used to raise the IEEE 754 inexact flag.
+         * In Java, there is no standard way to access these flags,
+         * so this check has no practical effect(?) and commented out.
+         */
+        // if (huge + x > 0.0) {
+        if (highWord >= 0) {
+          0.0
+        } else if (((highWord & 0x7fffffff) | lowWord) != 0) {
+          -1.0
+        } else { // -0.0
+          x
+        }
+        // }
+      } else { // Case 1b: 0 <= exponent < 20
+        val fractionalMask = 0x000fffff >>> exponent
+        if (((highWord & fractionalMask) | lowWord) == 0) { // x is integral
+          x
+        } else {
+          // if (huge + x > 0.0) { // Raise inexact flag
+          val adjustedHighWord = if (highWord < 0) {
+            // increment the integer part (and then clear the fractional part)
+            // 0x00100000 is a 21th bit
+            highWord + (0x00100000 >>> exponent)
+          } else {
+            highWord
+          }
+          // Clear fractional bits
+          reconstructDouble(adjustedHighWord & ~fractionalMask, 0)
+        }
+      }
+    } else if (exponent > 51) { // Case 2: Very large numbers or special values
+      if (exponent == 0x400) { // Infinity or NaN
+        x + x // returns qNaN
+      } else {
+        x
+      }
+    } else { // Case 3: 20 <= exponent <= 51: decimal point is in the low 32 bits
+      val fractionalMask = (0xffffffff >>> (exponent - 20)).toInt // fractional bits in low word
+      if ((lowWord & fractionalMask) == 0) { // x is integer
+        x
+      } else {
+        // if (huge + x > 0.0) {
+        val (adjustedHighWord, adjustedLowWord) =
+          if (highWord < 0) {
+            if (exponent == 20) {
+              (highWord + 1, lowWord)
+            } else {
+              val newLowWord = lowWord + (1 << (52 - exponent)) // + LSB of integer part
+              if (Integer.compareUnsigned(newLowWord, lowWord) < 0) { // carry
```

Comment:
```suggestion
              if (Integer.unsigned_<(newLowWord, lowWord)) { // carry
```
(but you won't need that with a `Long`-based implementation at all)

Reply:
Yup, we no longer need carry with a Long-based implementation.